src/shared/image_builder_lib.sh

```diff
@@ -0,0 +1,299 @@
+
+  declare -a additional_tags=("${CICD_TOOLS_IMAGE_BUILDER_ADDITIONAL_TAGS[@]:-${ADDITIONAL_TAGS[@]}}")
+
+  if cicd_tools::image_builder::_array_empty "${additional_tags[@]}"; then
+  additional_tags=()
+  fi
+
+  echo -n "${additional_tags[@]}"
+}
+
+cicd_tools::image_builder::_array_empty() {
+  local arr=("$1")
+
+  [[ "${#arr[@]}" -eq 1 && -z "${arr[0]}" ]]
+}
+
+cicd_tools::image_builder::get_labels() {
+
+  declare -a labels=("${CICD_TOOLS_IMAGE_BUILDER_LABELS[@]:-${LABELS[@]}}")
+
+  if cicd_tools::image_builder::_array_empty "${labels[@]}"; then
+  labels=()
+  fi
+
+  if cicd_tools::image_builder::is_change_request_context; then
+  labels+=("$(cicd_tools::image_builder::_get_expiry_label)")
+  fi
+
+  echo -n "${labels[@]}"
+}
+
+cicd_tools::image_builder::_get_expiry_label() {
+  echo "quay.expires-after=${CICD_TOOLS_IMAGE_BUILDER_QUAY_EXPIRE_TIME}"
+}
+
+cicd_tools::image_builder::get_build_args() {
+
+  declare -a build_args=("${CICD_TOOLS_IMAGE_BUILDER_BUILD_ARGS[@]:-${BUILD_ARGS[@]}}")
+
+  if cicd_tools::image_builder::_array_empty "${build_args[@]}"; then
+  build_args=()
+  fi
+
+  echo -n "${build_args[@]}"
+}
+
+cicd_tools::image_builder::tag() {
+
+  local source_image target_tag image_name
+  image_name="$(cicd_tools::image_builder::_get_image_name)" || return 1
+  source_image="$(cicd_tools::image_builder::get_full_image_name)" || return 1
+
+  for target_tag in $(cicd_tools::image_builder::get_additional_tags); do
+    if ! cicd_tools::container::cmd tag "$source_image" "${image_name}:${target_tag}"; then
+        cicd_tools::err "Error tagging '$source_image' as '${image_name}:${target_tag}'"
+        return 1
+    fi
+  done
+}
+
+cicd_tools::image_builder::push() {
+
+  local image_name image_tag
+  image_name="$(cicd_tools::image_builder::_get_image_name)" || return 1
+  image_tag=$(cicd_tools::image_builder::get_image_tag) || return 1
+
+  image_tags=("$image_tag")
+
+  if ! cicd_tools::image_builder::is_change_request_context; then
+  for additional_tag in $(cicd_tools::image_builder::get_additional_tags); do
+    image_tags+=("${additional_tag}")
+  done
+  fi
+
+  for tag in "${image_tags[@]}"; do
+  if ! cicd_tools::container::cmd push "${image_name}:${tag}"; then
+    cicd_tools::err "Error pushing image: '${image_name}:${tag}'"
+    return 1
+  fi
+  done
+}
+
+cicd_tools::image_builder::get_full_image_name() {
+
+  local image_name image_tag
+  image_name="$(cicd_tools::image_builder::_get_image_name)" || return 1
+  image_tag=$(cicd_tools::image_builder::get_image_tag) || return 1
+
+  echo -n "${image_name}:${image_tag}"
+}
+
+
+cicd_tools::image_builder::_image_builder_setup() {
+
+  if ! cicd_tools::image_builder::_try_log_in_to_image_registries; then
+    cicd_tools::err "Error trying to log into the image registries!"
+    return 1
+  fi
+}
+
+cicd_tools::image_builder::_try_log_in_to_image_registries() {
+
+  if cicd_tools::image_builder::_quay_credentials_found; then
```

Comment:
If creds are not found, should we print an error indicating that?

Reply:
> So the way I designed this behavior was, just to log in to a container registry if credentials were found on the environment. This will happen in CI - assuming the CI system is correctly configured - and probably won't happen locally, as the user probably won't have registry credentials exported on their shell by default.

OK, I get what you are saying about the local builkd. Yes, I was thinking about the case of the CI system being incorrectly configured. A pipeline job could be created that accidentally does not have the env vars set on it. Just consider how these env vars wind up on our jobs when using app-interface's JJB configs: you need to have the job pointing to the right job template, and then the job template must have a line that references the correct vault secret, and then you need to have the vault secret correctly defined. It's easy to mess up if you do not use the `insights` job templates. So if these scripts started with neither of these env vars passed to it, at the moment we would just silently continue right?

Reply:
Exactly, that would be the case today. And if there's no secret configured, it will just "try" and if it succeeds (because the process has something "cached" then it won't complain). 

One of the things I considered was to force the removal of the DOCKER_CONF, kind of similar to what we do in the old bootstrap scripts - but only in `CI` context (to not to mess up with local configs and whatnot). 

What I was thinking of a good middle ground was: 
- On CI context, always create a temporary `DOCKER_CONF` (that can be anywhere really, and we can leverage `mktemp` for it)
- don't change behavior - if no one configured a secret for the pipeline, no secret will be reused, so if no pushes are done , and no pulls from private regs. are done, then ... it wouldn't fail

WDYT? would it make sense ?

Reply:
The current scripts have a check here: https://github.com/RedHatInsights/cicd-tools/blob/main/_common_container_logic.sh#L37-L45

It is at the very bottom below a bunch of function definitions, so it could be easy to miss ... that code actually runs as soon as `_common_container_logic` is sourced. At the moment, we expect both of those env vars to always be there. I know the way you are re-writing things, you are supporting 4 use cases:
* only docker available
* only podman available
* both are there, and we prefer one
* "local" mode -- not running in CI

So I think somehow, I would prefer that we still error out as soon as we can if we see the creds to log in are not set on the CI job -- definitely when running on a CI system at least. For local mode, maybe printing a log message that we are "Running in local mode, using the current docker/podman context" is good?

What if local mode is controlled by an env var? Kind of like setting `DEBUG=true` what if we require people to set `CICD_LOCAL=true`? So the assumption can be we are always running on a CI system *unless* CICD_LOCAL is true. That could make it easy to code the logic in for this kind of stuff. Or do you think it would be too annoying to require devs to set that env var when they want to debug the scripts locally?

By the way I think when running on CI, creating a temporary docker conf using mktemp would be a good way to do it :)

Reply:
Thanks for the feedback, my comments follow semi-inline: 

> know the way you are re-writing things, you are supporting 4 use cases
You meant ... the RH registry and the Quay registry?  seems like it all got mixed with the docker/podman detection. In away, yes, it applies for all of those options.

> what if we require people to set CICD_LOCAL=true?
Well this was exactly the actual initial implementation I went with I started the work on the compliance-backend library: https://github.com/RedHatInsights/compliance-backend/blob/7c2bc8a763c794505b32cae501b1f5611961fceb/scripts/deploy/build-deploy-common.sh#L20-L22

And yeah, it was very annoying. I thought about trying to be smart enough and autodetect this with the "CI" variable, and with a similar approach with the "MR and GH variables" to detect if on a change request context. 

I do like much more this approach, and the fact is that I kept the local_build flag to override this if needed. so - one would want to "test" with CI=true and yet force "local mode" . This is what I mean: https://github.com/RedHatInsights/cicd-tools/blob/a7b807231558e06ed7e32919cf2952421b12c41b/src/shared/common_lib.sh#L36-L38

I just found that I wouldn't have to do any of this by using `is_ci_context` all alone. but maybe I can use that instead of just relying on the CI context.

>By the way I think when running on CI, creating a temporary docker conf using mktemp would be a good way to do it :)

Already implementing this :)

Reply:
@bsquizz  I just forgot, but, the library already fails if it's supposed to login to the registry and can't (the case where, the credentials are set, but the login fails somehow). 

Anyways - I've made a few changes related to this very last bit:
- I've changed the condition of "not pushing" on the "build and push" function to "if not a local build" . That means, either it's not CI=true or is not explicitly forced local build with exporting the `LOCAL_BUILD` variable to a non null value.
- I've also forced fresh registry credentials when on a non-local build context, which uses `mktemp`

Reply:
OK, I am still a little confused on what happens if the registry credential env vars are totally missing and we run this on a CI system. Don't we need a line like this: "if not local build and not quay_credentials_found and not redhat_registry_credentials_found, return 1" defined in this func: https://github.com/RedHatInsights/cicd-tools/pull/22/files#diff-115966db1a9e3d141663dc0c5927fdfb3f045d8ffc2662996d0b01151ddcae11R243 ?

Reply:
OK I guess the end result here is, the job will just fail when it tries to pull an image from a registry it is not logged in to. Or, it'll fail at the end when it tries to run the push because it won't be logged in. Since we create the isolated docker config, that at least prevents us from being accidentally logged in using other creds